### PR TITLE
[CA-928, CA-968] Add descendant permissions to policies

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [CA-928](https://broadworkbench.atlassian.net/browse/CA-928), [CA-968](https://broadworkbench.atlassian.net/browse/CA-928)
Add descendant permissions to policies so that policies can grant permissions to child resources of a given resource type
Bump Resource routes to v2

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyTable.scala [hunk 1]
```diff
@@ -20,6 +21,7 @@ object PolicyTable extends SQLSyntaxSupportWithDefaultSamDB[PolicyRecord] {
     rs.get(e.resourceId),
     rs.get(e.groupId),
     rs.get(e.name),
-    rs.get(e.public)
+    rs.get(e.public),
+    rs.longOpt(e.descendantResourceTypeId).map(ResourceTypePK)
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyTable.scala]: I forgot to update Postgres to actually write this value, but now I'm not sure I even understand this. What if a policy has descendant permissions that apply to multiple different kinds of resource types? Which `ResourceTypePK` would go here?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 2]
```diff
@@ -322,6 +324,8 @@ class ResourceService(
       validateRoles(resourceType, policy) ++
       validateUrlSafe(policy.policyName.value)
 
+    // todo: validate descendant permissions??
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: As far as I can tell, I do need to add something here to make sure that the policy isn't inherited by any other resources and to throw a 400 if it is

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyTable.scala (hunk 1 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyTable.scala]: My current thinking is that each role and action in the `PolicyRole` and `PolicyAction` tables has to be associated with a resource type, so there's no need to define which resource type any given policy might apply to using this FK. If there are roles or actions defined in the `PolicyRole` or `PolicyAction` table that descend, then those roles and actions apply to child resources of their given resource type.

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200827_descendant_permissions.xml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="dummy"
+                   xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+    <changeSet logicalFilePath="dummy" author="mtalbott" id="descendant_permissions">
```
Comment (dvoet) [inline src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200827_descendant_permissions.xml]: Put each add column in a different change set. This allows each change set to be atomic and avoid problems where the change set fails part way through and is applied incompletely. This is important for DDL statements since they don't participate in transaction and thus have no rollback.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyTable.scala (hunk 1 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyTable.scala]: yes!

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 4]
```diff
@@ -124,11 +124,11 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
...
   }
 
-  private def insertActions(actions: Set[ResourceAction], resourceTypePK: ResourceTypePK)(implicit session: DBSession): Int = {
+  private def insertActions(actions: Set[(ResourceAction, ResourceTypePK)])(implicit session: DBSession): Int = {
     if (actions.isEmpty) {
-      return 0
+      0
     } else {
-      val uniqueActionValues = actions.map { action =>
+      val uniqueActionValues = actions.collect { case (action, resourceTypePK) =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: this should still be map, not collect

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 5]
```diff
@@ -353,7 +353,9 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
       insertGroupMembers(groupId, policy.members)
 
       insertPolicyRoles(policy.roles, policyId)
+      insertDescendantRoles(policy.descendantPermissions.flatMap(permissions => permissions.roles.map((_, permissions.resourceType))), policyId)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: I think we can have a single function instead of 2: `insertPolicyRoles` and `insertDescendantRoles`. The only real difference is that one inserts true for `descends` and the other false. Same goes for actions.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 6]
```diff
@@ -362,6 +364,60 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
     }
   }
 
+  private def insertDescendantRoles(rolesWithResourceType: Set[(ResourceRoleName, ResourceTypeName)], policyId: PolicyPK)(implicit session: DBSession): Int = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: I feel like `(ResourceRoleName, ResourceTypeName)` should be its own type: 
```
case class FullyQualifiedResourceRoleName(name: ResourceRoleName, resourceTypeName: ResourceTypeName)
```

maybe a companion object:
```
object FullyQualifiedResourceRoleName {
  def fullyQualify(roles: Set[ResourceRoleName], resourceTypeName: ResourceTypeName) = roles.map(FullyQualifiedResourceRoleName(_, resourceTypeName)
}
```

Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: I tried to work out a way to combine this with `insertPolicyRoles` (and to combine `insertDescendantActions` with `insertPolicyActions`), but I couldn't work out how to do it. If anyone has any ideas, let me know!

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 7]
```diff
@@ -372,6 +387,29 @@ class ResourceService(
     } else None
   }
 
+  private def validatePolicyNotInherited(policy: FullyQualifiedPolicyId, samRequestContext: SamRequestContext): IO[Option[ErrorReport]] = {
+    for {
+      children <- accessPolicyDAO.listResourceChildren(policy.resource, samRequestContext)
+      policyOpt <- accessPolicyDAO.loadPolicy(policy, samRequestContext)
+    } yield {
+      policyOpt.flatMap { policy =>
+        if (policyInheritedByChildren(policy.descendantPermissions, children)) {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: why are we preventing this case?

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200901_descendant_policy_roles.xml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="dummy"
+                   xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+    <changeSet logicalFilePath="dummy" author="mtalbott" id="descendant_policy_roles">
```
Comment (dvoet) [inline src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200901_descendant_policy_roles.xml]: you can have more than one `changeSet` in a `databaseChangeLog`

Comment (dvoet) [reply src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200901_descendant_policy_roles.xml]: in other words, you don't need 2 files

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 7 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: According to the ticket, overwriting the policy should "Return 400 if policy is inherited."

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200901_descendant_policy_roles.xml (hunk 8 above)
Comment (marctalbott) [reply src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200901_descendant_policy_roles.xml]: yeah, I figured this out when I was making some other liquibase changes, it's back to one file in future commits

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 7 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: @dvoet is this still something we want to prevent? Or was the ticket description a relic of when you could reference inherited policies with a '~'?

File: docker/run-opendj.sh [hunk 9]
```diff
@@ -10,6 +10,10 @@ start()  {
             CONTAINER_NAME=opendj-${BUILD_TAG}
         fi
     fi
+
+    echo "attempting to remove old $CONTAINER_NAME container..."
```
Comment (marctalbott) [inline docker/run-opendj.sh:14]: This is an unrelated QoL change that the run-postgres.sh script has

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 7 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: Scrapped this as the ticket description was out of date

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala [hunk 10]
```diff
@@ -236,41 +329,41 @@ trait ResourceRoutes extends UserInfoDirectives with SecurityDirectives with Sam
...
-        entity(as[AccessPolicyMembership]) { membershipUpdate =>
+      entity(as[AccessPolicyMembership]) { membershipUpdate =>
+        requireAction(policyId.resource, SamResourceActions.alterPolicies, userInfo.userId, samRequestContext) {
           complete(resourceService.overwritePolicy(resourceType, policyId.accessPolicyName, policyId.resource, membershipUpdate, samRequestContext).map(_ => StatusCodes.Created))
         }
       }
     }
+  }
+
+  private def requireActionsForSharePolicy(policyId: FullyQualifiedPolicyId, userInfo: UserInfo, samRequestContext: SamRequestContext)(sharePolicy: server.Route): server.Route =
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala:342]: Previously we were doing the same permissions check in two different places in the v1 routes. I've consolidated that check into a single place so that it will be the same check for all three endpoints in both versions

File: src/main/resources/swagger/api-docs.yaml [hunk 11]
```diff
@@ -2009,6 +2774,17 @@ components:
         policy:
           $ref: '#/components/schemas/AccessPolicyMembership'
       description: ""
+    AccessPolicyResponseEntryV2:
```
Comment (marctalbott) [inline src/main/resources/swagger/api-docs.yaml:2777]: I opted to make this a separate component instead of just adding descendant permissions as an optional property so that they wouldn't show up at all in the v1 payloads in swagger in an effort to be conscious of the swagger generated client by not changing the v1 endpoints at all

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/model/SamModel.scala [hunk 12]
```diff
@@ -156,10 +158,18 @@ consistent "has a" relationship is tracked by this ticket: https://broadworkbenc
...
     public: Boolean)
     extends WorkbenchGroup
 
-@Lenses final case class AccessPolicyMembership(memberEmails: Set[WorkbenchEmail], actions: Set[ResourceAction], roles: Set[ResourceRoleName])
+@Lenses final case class AccessPolicyDescendantPermissions(resourceType: ResourceTypeName, actions: Set[ResourceAction], roles: Set[ResourceRoleName])
+@Lenses final case class AccessPolicyMembership(memberEmails: Set[WorkbenchEmail],
+                                                actions: Set[ResourceAction],
+                                                roles: Set[ResourceRoleName],
+                                                descendantPermissions: Option[Set[AccessPolicyDescendantPermissions]] = None) {
+  def toV1 = this.copy(descendantPermissions = None) // strip out descendantPermissions for backwards compatibility
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/model/SamModel.scala]: I don't think we need to do this. Adding new fields to our response payloads is perfectly backwards compatible. It is also backwards compatible for requests since it is an `Option` and thus not required.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala [hunk 13]
```diff
@@ -236,41 +329,41 @@ trait ResourceRoutes extends UserInfoDirectives with SecurityDirectives with Sam
...
-  def getResourcePolicies(resource: FullyQualifiedResourceId, userInfo: UserInfo, samRequestContext: SamRequestContext): server.Route =
+  private def requireActionsForListPolicies(resource: FullyQualifiedResourceId, userInfo: UserInfo, samRequestContext: SamRequestContext)(listResourcePolicies: server.Route): server.Route =
     get {
       requireAction(resource, SamResourceActions.readPolicies, userInfo.userId, samRequestContext) {
-        complete(resourceService.listResourcePolicies(resource, samRequestContext).map { response =>
-          StatusCodes.OK -> response.toSet
-        })
+        listResourcePolicies
       }
     }
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala]: this would be more akka http friendly if written as a directive:
```
  private def requireActionsForListPolicies(resource: FullyQualifiedResourceId, userInfo: UserInfo, samRequestContext: SamRequestContext): Directive0 =
    get & requireAction(resource, SamResourceActions.readPolicies, userInfo.userId, samRequestContext)
```
BUT I would not include the `get` in there. With the name `requireActionsForListPolicies`, including the `get` is unexpected.

**Review by dvoet — CHANGES_REQUESTED** [review]
I think there is too much change going on the the v1 resource routes. I think the v1 routes should be unchanged. The process should be copy all the v1 routes to v2, add in the new v2 routes. It is ok to change the underlying functions (like `getResourcePolicies`) because they are backwards compatible. The reason to the version bump so far is more of a semantic change instead of a break in compatibility (although that break is coming soon).

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala (hunk 10 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala:342]: This is no longer used in the v1 routes because we don't want to change the v1 routes at all in this PR. I'd prefer to keep it for the v2 routes at least, but can also remove this optimization and leave it entirely for a separate PR

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/model/SamModel.scala [hunk 14]
```diff
@@ -156,10 +158,17 @@ consistent "has a" relationship is tracked by this ticket: https://broadworkbenc
...
+    descendantPermissions: Set[AccessPolicyDescendantPermissions],
     public: Boolean)
     extends WorkbenchGroup
 
-@Lenses final case class AccessPolicyMembership(memberEmails: Set[WorkbenchEmail], actions: Set[ResourceAction], roles: Set[ResourceRoleName])
+@Lenses final case class AccessPolicyDescendantPermissions(resourceType: ResourceTypeName, actions: Set[ResourceAction], roles: Set[ResourceRoleName])
+@Lenses final case class AccessPolicyMembership(memberEmails: Set[WorkbenchEmail],
+                                                actions: Set[ResourceAction],
+                                                roles: Set[ResourceRoleName],
+                                                descendantPermissions: Option[Set[AccessPolicyDescendantPermissions]] = Option(Set.empty)) {
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/model/SamModel.scala:169]: I changed the default value from `None` to `Option(Set.empty)` so that I wouldn't have to change the v1 routes specs at all. With `None` as the default value, the expected response would be `AccessPolicyMembership(_, _, _, None)` but the response returned by the ResourceService is `AccessPolicyMembership(_, _, _, Some(Set.empty))`

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200901_descendant_permissions.xml [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+    <changeSet logicalFilePath="dummy" author="mtalbott" id="drop_policy_action_PK">
+        <dropPrimaryKey tableName="sam_policy_action" constraintName="sam_policy_action_pkey"/>
+    </changeSet>
+
+    <changeSet logicalFilePath="dummy" author="mtalbott" id="descendant_policy_actions">
+        <addColumn tableName="sam_policy_action">
+            <column name="descends" type="BOOLEAN" defaultValue="false">
```
Comment (andy7i) [inline src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200901_descendant_permissions.xml]: `descends` makes me think that this permission would start at the parent and then apply to descendants. I don't have a great alternative but maybe couple options: `descendantsOnly` or `appliesToAllDescendantsOnly`?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyActionTable.scala [hunk 16]
```diff
@@ -4,14 +4,16 @@ import scalikejdbc._
 import org.broadinstitute.dsde.workbench.sam.db.SamTypeBinders
 
 final case class PolicyActionRecord(resourcePolicyId: PolicyPK,
-                                    resourceActionId: ResourceActionPK)
+                                    resourceActionId: ResourceActionPK,
+                                    descends: Boolean)
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyActionTable.scala]: same comment as above:

> `descends` makes me think that this permission would start at the parent and then apply to descendants. I don't have a great alternative but maybe couple options: `descendantsOnly` or `appliesToAllDescendantsOnly`?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyRoleTable.scala [hunk 17]
```diff
@@ -4,14 +4,16 @@ import scalikejdbc._
 import org.broadinstitute.dsde.workbench.sam.db.SamTypeBinders
 
 final case class PolicyRoleRecord(resourcePolicyId: PolicyPK,
-                                  resourceRoleId: ResourceRolePK)
+                                  resourceRoleId: ResourceRolePK,
+                                  descends: Boolean)
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/db/tables/PolicyRoleTable.scala]: same comment as above:

> `descends` makes me think that this permission would start at the parent and then apply to descendants. I don't have a great alternative but maybe couple options: `descendantsOnly` or `appliesToAllDescendantsOnly`?

**Review by dvoet — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 18]
```diff
@@ -362,21 +364,19 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
     }
   }
 
-  private def insertPolicyActions(actions: Set[ResourceAction], policyId: PolicyPK)(implicit session: DBSession): Int = {
+  private def insertPolicyActions(actions: Set[FullyQualifiedResourceAction], policyId: PolicyPK, descends: Boolean)(implicit session: DBSession): Int = {
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: same comment as above, also applies a couple times in this file:

> `descends` makes me think that this permission would start at the parent and then apply to descendants. I don't have a great alternative but maybe couple options: `descendantsOnly` or `appliesToAllDescendantsOnly`?

**Review by andy7i — COMMENTED** [review]
adding comments so far, which are all about the naming of `descends` (one comment per file)

(I still need to finish reviewing DAO changes, and then review swagger and tests.)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 19]
```diff
@@ -362,21 +364,19 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
...
+        case FullyQualifiedResourceAction(action, resourceType) => samsqls"(${action}, ${resourceType})"
+      }
+      val insertQuery = samsqls"""insert into ${PolicyActionTable.table} (${paCol.resourcePolicyId}, ${paCol.resourceActionId}, ${paCol.descends})
+            select ${policyId}, ${ra.result.id}, ${descends}
             from ${ResourceActionTable as ra}
             join ${ResourceTypeTable as rt} on ${ra.resourceTypeId} = ${rt.id}
-            join ${ResourceTable as r} on ${r.resourceTypeId} = ${rt.id}
-            join ${PolicyTable as p} on ${p.resourceId} = ${r.id}
-            where ${ra.action} in (${actions})
-            and ${p.id} = ${policyId}"""
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala:379]: Just to double check that I understand this: the old code here verified that the `policyId` exists. This was removed since this method takes in a `policyId: PolicyPK` which we can assume is valid and exists?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 18 above)
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: not related to this PR: this function returns an Int, but it doesn't seem to be used anywhere. Do we return the int because it's easy to return that data and might be useful for future use?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 20]
```diff
@@ -385,9 +385,16 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
...
-          .map(rs => rs.get[ResourceTypePK](r.resultName.resourceTypeId)).single().apply().getOrElse(throw new WorkbenchException(s"could not find resource type id for policy id $policyId"))
-        insertActions(actions, resourceTypePK)
+        val resourceTypeNames = actions.map(_.resourceTypeName)
+
+        val resourceTypeNameAndPK = samsql"select ${rt.result.name}, ${rt.result.id} from ${ResourceTypeTable as rt} where (${rt.name}) in (${resourceTypeNames})"
+          .map(rs => (rs.get[ResourceTypeName](rt.resultName.name), rs.get[ResourceTypePK](rt.resultName.id))).list().apply().toMap
+        val actionsWithResourceTypePK = actions.map {
+          case FullyQualifiedResourceAction(action, resourceTypeName) =>
+            (action, resourceTypeNameAndPK.getOrElse(resourceTypeName,
+              throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, "Resource type not found"))))
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: would it be helpful to the user to include the resource type name in the error message?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 21]
```diff
@@ -611,34 +620,61 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
...
+
+    val descendantPermissions = descendantRoleResults.groupBy(_.resourceTypeName).collect { case (Some(descendantResourceTypeName), descendantRolesWithResourceType) =>
+      val descendantRoles = descendantRolesWithResourceType.collect { case RoleResult(_, Some(role), _) => role }
+      val descendantActions = descendantActionResults.groupBy(_.resourceTypeName).getOrElse(Option(descendantResourceTypeName), List.empty).collect { case ActionResult(_, Some(action), _) => action }
+      AccessPolicyDescendantPermissions(descendantResourceTypeName, descendantActions.toSet, descendantRoles.toSet)
+    }.toSet
+
+    (roles, actions, descendantPermissions)
+  }
+
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: nice 👍

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 22]
```diff
@@ -920,6 +956,19 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
 }
 
 private final case class PolicyInfo(name: AccessPolicyName, resourceId: ResourceId, resourceTypeName: ResourceTypeName, email: WorkbenchEmail, public: Boolean)
+private final case class MemberResult(userId: Option[WorkbenchUserId], groupName: Option[WorkbenchGroupName], policyName: Option[AccessPolicyName], resourceId: Option[ResourceId], resourceTypeName: Option[ResourceTypeName])
+private final case class RoleResult(resourceTypeName: Option[ResourceTypeName], role: Option[ResourceRoleName], descends: Option[Boolean])
+private final case class ActionResult(resourceTypeName: Option[ResourceTypeName], action: Option[ResourceAction], descends: Option[Boolean])
+
+final case class FullyQualifiedResourceRoleName(roleName: ResourceRoleName, resourceTypeName: ResourceTypeName)
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: would `FullyQualifiedResourceRole` be a better name?

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAOSpec.scala [hunk 23]
```diff
@@ -499,10 +499,31 @@ class PostgresAccessPolicyDAOSpec extends FreeSpec with Matchers with BeforeAndA
         dirDao.createGroup(defaultGroup, samRequestContext = samRequestContext).unsafeRunSync()
         dirDao.createUser(defaultUser, samRequestContext).unsafeRunSync()
 
-        val policy = AccessPolicy(FullyQualifiedPolicyId(resource.fullyQualifiedId, AccessPolicyName("policyName")), Set(defaultGroup.id, defaultUser.id), WorkbenchEmail("policy@email.com"), resourceType.roles.map(_.roleName), Set(readAction, writeAction), false)
+        val policy = AccessPolicy(FullyQualifiedPolicyId(resource.fullyQualifiedId, AccessPolicyName("policyName")), Set(defaultGroup.id, defaultUser.id), WorkbenchEmail("policy@email.com"), resourceType.roles.map(_.roleName), Set(readAction, writeAction), Set.empty, false)
         dao.createPolicy(policy, samRequestContext).unsafeRunSync()
         dao.loadPolicy(policy.id, samRequestContext).unsafeRunSync() shouldEqual Option(policy)
       }
+
+      "creates descendant actions and roles" in {
```
Comment (andy7i) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAOSpec.scala:507]: would it make sense to add another test where we verify that a "descends" policy cannot be used by the parent resource?

**Review by andy7i — APPROVED** [review]
minor questions, nothing blocking. 👍

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 19 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala:379]: Not exactly. Resource actions are resource type specific. For example, you could have two different `write` resource actions, one for the `workspace` resource type and another for the `billing-project` resource type. When we insert into the `PolicyActionTable`, we need to make sure we are inserting the correct action. 

Before this PR, the correct action was always on the same resource type as the policy. Given that, we would select `${ra.result.id}` based on the resource type of the policy -- that's also why we needed the additional joins to the `ResourceTable` and the `PolicyTable`. 

Now we want to be able to insert actions for arbitrary resource types, not just the resource type of the policy. We are able to do this by passing in the resource type name along with the action and then comparing them to the `(${ra.action}, ${rt.name})` tuples in the conditional statement on l. 379.

So you're right in that we can assume that the `policyId` exists and is valid because we've previously loaded that from Postgres, but that's not why we don't need the conditional anymore

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 18 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: Yup, by default the `.apply()` returns the number of rows updated, so we just return that number instead of dropping it. I'm not sure if we ever really use it to be honest, but it's good practice to return it anyways I think

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 22 above)
Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: I think I do like that better. I went back and forth on that a bit and settled on what I have because it matched with `ResourceRoleName`, but I don't think it has to match and your suggestion rolls off the tongue better

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAOSpec.scala (hunk 23 above)
Comment (marctalbott) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAOSpec.scala:507]: Discussed on slack, Doug's PR will include a test to this effect
